src/repr/type_binary.ml

```diff
@@ -19,112 +19,110 @@ open Staging
 open Utils
 
 module Encode = struct
-  let chars =
-    Array.init 256 (fun i -> Bytes.unsafe_to_string (Bytes.make 1 (Char.chr i)))
+  let unit () _byt off = off
 
-  let unit () _k = ()
-  let unsafe_add_bytes b k = k (Bytes.unsafe_to_string b)
-  let add_string s k = k s
-  let char c k = k chars.(Char.code c)
+  let add_string s byt off =
+    let ls = String.length s in
+    Bytes.blit_string s 0 byt off ls;
+    off + ls
 
-  let int8 i k =
-    assert (i < 256);
-    k chars.(i)
+  let add_bytes b byt off =
+    let lb = Bytes.length b in
+    Bytes.blit b 0 byt off lb;
+    off + lb
 
-  let int16 i =
-    let b = Bytes.create 2 in
-    Bytes.set_uint16_be b 0 i;
-    unsafe_add_bytes b
+  let char c byt off =
+    Bytes.set byt off c;
+    off + 1
 
-  let int32 i =
-    let b = Bytes.create 4 in
-    Bytes.set_int32_be b 0 i;
-    unsafe_add_bytes b
+  let byte n byt off = char (Char.chr (0xFF land n)) byt off
```

Comment:
I think I missed something. why are you doing `0xFF land n`

Reply:
A lot of the things you noticed were things I forgot to clean, thanks and sorry for making you do this :)